src/GafferScene/ImageSampler.cpp

```diff
@@ -0,0 +1,426 @@
+		}
+	}
+}
+
+IECore::ConstObjectPtr ImageSampler::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, const IECore::Object *inputObject ) const
+{
+	const Primitive *primitive = runTimeCast<const Primitive>( inputObject );
+	if( !primitive )
+	{
+		return inputObject;
+	}
+	
+	PrimitivePtr outputPrimitive = primitive->copy();
+
+	const std::string channelNames = channelsPlug()->getValue();
+	const std::string primVarName = primVarNamePlug()->getValue();
+	const std::string uvVarName = uvVarNamePlug()->getValue();
+
+	if( channelNames.empty() || primVarName.empty() || uvVarName.empty() )
+	{
+		return inputObject;
+	}
+
+	const auto uvVarNameIt = outputPrimitive->variables.find( uvVarName );
+	if( uvVarNameIt == outputPrimitive->variables.end() )
+	{
+		IECore::msg(IECore::Msg::Warning, "ImageSampler::computeProcessedObject", boost::format( "UV primitive variable \"%s\" does not exist." ) % uvVarName.c_str());
+		return inputObject;
+	}
+
+	PrimitiveVariable::IndexedView<V2f> uvView( uvVarNameIt->second );
+	const PrimitiveVariable::Interpolation interpolation = uvVarNameIt->second.interpolation;
+
+	ConstStringVectorDataPtr inChannelNamesData = imagePlug()->channelNamesPlug()->getValue();
+	const std::vector<std::string> inChannelNames = inChannelNamesData->readable();
+
+	const Imath::V2i imageSize = imagePlug()->dataWindow().size();
+	const Imath::V2i imageMin = imagePlug()->dataWindow().min;
+
+	std::vector<std::string> channels;
+	IECore::StringAlgo::tokenize( channelNames, ' ', channels );
+
+	// TODO: Check for missing channels?
+
+	if( ( primVarName == "Cs" && channels.size() != 3 ) ||
+		( primVarName == "N" && channels.size() != 3 ) ||
+		( primVarName == "P" && channels.size() != 3 ) ||
+		( primVarName == "Pref" && channels.size() != 3 ) ||
+		( primVarName == "scale" && channels.size() != 3 ) ||
+		( primVarName == "velocity" && channels.size() != 3 )
+		
+	)
+	{
+		throw IECore::Exception( "Primitive variable \"" + primVarName + "\" must sample three channels." );
+	}
+
+	if( primVarName == "uv" && channels.size() != 2 )
+	{
+		throw IECore::Exception( "Primitive variable \"uv\" must sample two channels." );
+	}
+
+	if( primVarName == "width" && channels.size() != 1)
+	{
+		throw IECore::Exception( "Primitive variable \"width\" must sample one channels." );
+	}
+
+	std::vector<OutputVariableFunction> outputs;
+
+	if( primVarName == "Cs" ||
+		primVarName == "N" ||
+		primVarName == "P" ||
+		primVarName == "Pref" ||
+		primVarName == "scale" ||
+		primVarName == "uv" ||
+		primVarName == "velocity" ||
+		primVarName == "width"
+	)
+	{
+		for( size_t i = 0; i < channels.size(); ++i )
+		{
+			if( auto o = addPrimitiveVariable( outputPrimitive.get(), primVarName, interpolation ) )
+			{
+				outputs.push_back( o );
+			}
+		}
+	}
+	else
+	{
+		for( size_t i = 0; i < channels.size(); ++i )
+		{
+			if( auto o = addPrimitiveVariable( outputPrimitive.get(), primVarName + "." + channels[i], interpolation ) )
```

Comment:
I found this a bit confusing, particularly in the PrimitiveInspector where the multiple `prefix.suffix` primvars appear alongside single vector primvars which are labelled as `prefix.suffix` on a per-column basis. Is it especially useful to be able to sample multiple float channels when we can't sample multiple V3f/V2d channels etc? Is there merit in a simpler "one node, one primvar" rule?